modules/apps/frontend-js/frontend-js-web/src/main/java/com/liferay/frontend/js/web/internal/servlet/taglib/aui/PortletDataRendererImpl.java

```diff
@@ -0,0 +1,168 @@
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.frontend.js.web.internal.servlet.taglib.aui;
+
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.servlet.taglib.aui.AMDRequire;
+import com.liferay.portal.kernel.servlet.taglib.aui.ESMImport;
+import com.liferay.portal.kernel.servlet.taglib.aui.PortletData;
+import com.liferay.portal.kernel.servlet.taglib.aui.PortletDataRenderer;
+
+import java.io.CharArrayWriter;
+import java.io.IOException;
+import java.io.Writer;
+
+import java.util.Collection;
+
+import org.osgi.service.component.annotations.Component;
+
+/**
+ * @author Iván Zaera Avellón
+ */
+@Component(immediate = true, service = PortletDataRenderer.class)
+public class PortletDataRendererImpl implements PortletDataRenderer {
+
+	@Override
+	public void writeTo(Writer writer, Collection<PortletData> portletDatas)
+		throws IOException {
+
+		PortletDataCollection portletDataCollection = new PortletDataCollection(
+			portletDatas);
+
+		if (portletDataCollection.isUseESM()) {
+			writer.write("<script type=\"module\">\n// <![CDATA[\n");
+		}
+		else {
+			writer.write("<script type=\"text/javascript\">\n// <![CDATA[\n");
+		}
+
+		Collection<ESMImport> esmImports =
+			portletDataCollection.getESMImports();
+
+		if (!esmImports.isEmpty()) {
+			_writeESMProlog(writer, esmImports);
+		}
+
+		if (!portletDataCollection.isUseESM()) {
+			writer.write("(function() {\n");
+		}
+
+		writer.write("var $ = AUI.$;var _ = AUI._;\n");
+
+		Collection<AMDRequire> amdRequires =
+			portletDataCollection.getAMDRequires();
+
+		if (!amdRequires.isEmpty()) {
+			_writeAMDProlog(writer, amdRequires);
+		}
+
+		Collection<String> auiModules = portletDataCollection.getAUIModules();
+
+		if (!auiModules.isEmpty()) {
+			_writeAUIProlog(writer, auiModules);
+		}
+
+		writer.write(portletDataCollection.getContent());
+
+		if (!auiModules.isEmpty()) {
+			_writeAUIEpilog(writer);
+		}
+
+		if (!amdRequires.isEmpty()) {
+			_writeAMDEpilog(writer);
+		}
+
+		if (!portletDataCollection.isUseESM()) {
+			writer.write("})();\n");
+		}
+
+		writer.write("\n// ]]>\n</script>");
+	}
+
+	private void _writeAMDEpilog(Writer writer) throws IOException {
+		writer.write("} catch (err) {\n");
+		writer.write("\tconsole.error(err);\n");
+		writer.write("}\n");
+
+		writer.write("});\n");
+	}
+
+	private void _writeAMDProlog(
+			Writer writer, Collection<AMDRequire> amdRequires)
+		throws IOException {
+
+		writer.write("Liferay.Loader.require(\n");
+
+		for (AMDRequire amdRequire : amdRequires) {
+			writer.write("  '");
```

Comment:
I think SF usually gets mad about using literal space and single quote, so you may need to use the constants we have for them

Reply:
Yeah, that's true if the `'` is used alone, but if you put it inside a longer string it's allowed. We have some exotic pooled strings like `StringPool.COMMA_AND_SPACE` but nothing like `TWO_SPACES_AND_APOSTROPHE` :sweat_smile: